test/Microsoft.ML.TestFrameworkCommon/CommonUtilities.cs

```diff
@@ -0,0 +1,366 @@
+        {
+            Assert(size == 0 || v1.Length == size);
+            Assert(size == 0 || v2.Length == size);
+            Assert(v1.Length == v2.Length);
+
+            var v1Values = v1.GetValues();
+            var v2Values = v2.GetValues();
+
+            if (v1.IsDense && v2.IsDense)
+            {
+                for (int i = 0; i < v1.Length; i++)
+                {
+                    var x1 = v1Values[i];
+                    var x2 = v2Values[i];
+                    if (!fn(i, x1, x2))
+                        return false;
+                }
+                return true;
+            }
+
+            Assert(!v1.IsDense || !v2.IsDense);
+            int iiv1 = 0;
+            int iiv2 = 0;
+            var v1Indices = v1.GetIndices();
+            var v2Indices = v2.GetIndices();
+            for (; ; )
+            {
+                int iv1 = v1.IsDense ? iiv1 : iiv1 < v1Indices.Length ? v1Indices[iiv1] : v1.Length;
+                int iv2 = v2.IsDense ? iiv2 : iiv2 < v2Indices.Length ? v2Indices[iiv2] : v2.Length;
+                T x1, x2;
+                int iv;
+                if (iv1 == iv2)
+                {
+                    if (iv1 == v1.Length)
+                        return true;
+                    x1 = v1Values[iiv1];
+                    x2 = v2Values[iiv2];
+                    iv = iv1;
+                    iiv1++;
+                    iiv2++;
+                }
+                else if (iv1 < iv2)
+                {
+                    x1 = v1Values[iiv1];
+                    x2 = default(T);
+                    iv = iv1;
+                    iiv1++;
+                }
+                else
+                {
+                    x1 = default(T);
+                    x2 = v2Values[iiv2];
+                    iv = iv2;
+                    iiv2++;
+                }
+                if (!fn(iv, x1, x2))
+                    return false;
+            }
+        }
+
+        public static bool EqualTypes(DataViewType type1, DataViewType type2, bool exactTypes)
+        {
+            AssertValue(type1);
+            AssertValue(type2);
+
+            return exactTypes ? type1.Equals(type2) : type1.SameSizeAndItemType(type2);
+        }
+
+        private static void Fail(string fmt, params object[] args)
+        {
+            Fail(false, fmt, args);
+        }
+
+        private static void Fail(bool relax, string fmt, params object[] args)
+        {
+            Log("*** Failure: " + fmt, args);
+        }
+
+        private static bool Failed()
+        {
+            return false;
+        }
+
+        private static void Log(string message, params object[] args)
+        {
+            Console.WriteLine(message, args);
+        }
+
+        private static bool Check(bool f, string msg)
+        {
+            if (!f)
+                Fail(msg);
+            return f;
+        }
+
+        [Conditional("DEBUG")]
```

Comment:
I don't understand why we have these methods. These are tests, you should use the `Assert` class in xunit. If any conditions fail, it should fail the test, no matter if this is DEBUG or not.

Reply:
This function is copied from Contracts Utilities class and this is also widely used in our test project like below, so are these also abuse and should be fixed?

protected bool EqualTypes(DataViewType type1, DataViewType type2, bool exactTypes)
        {
            Contracts.AssertValue(type1);
            Contracts.AssertValue(type2);

            return exactTypes ? type1.Equals(type2) : type1.SameSizeAndItemType(type2);
        }



---
In reply to: [337248885](https://github.com/dotnet/machinelearning/pull/4346#discussion_r337248885) [](ancestors = 337248885)

Reply:
Just `Assert.IsNotNull` instead of `Contracts.AssertValue`